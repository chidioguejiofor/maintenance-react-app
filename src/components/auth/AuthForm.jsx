import { Form, Container, Header, Icon, Message } from 'semantic-ui-react';
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import TopNav from '../Header';

const AuthForm = ({
  formType, handleSubmit, success, messageTitle,
  loading, handleChange, inputs, errorMessages,
  links
}) => (
  <div id="auth-form">
    <TopNav />
    <Container>
      <div className="form">
        <Form
          loading={loading}
          onSubmit={handleSubmit}
          error={!!errorMessages.length}
          success={success}
        >

          <Header as="h2" icon textAlign="center">
            <Icon name="user circle" />
            <Header.Content>{formType}</Header.Content>
          </Header>
          <Message
            header={messageTitle}
            error={!!errorMessages.length}
            success={success}
            list={errorMessages}
            hidden={!success && !errorMessages.length}
          />
          {inputs.map((input, index) => (
            <Form.Input
              key={index}
              label={input.placeholder}
              placeholder={input.placeholder}
              name={input.name}
              value={input.value}
              onChange={handleChange}
              type={input.type}
            />
          ))}
          <Form.Button
            disabled={success}
            positive
            content="Submit"
          />
          {links.map((linkObj, index) => (
            <span key={index} className="link">
              <Link to={linkObj.to}>{linkObj.caption}</Link>
            </span>

          ))}
        </Form>
      </div>

    </Container>
  </div>


);

AuthForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  inputs: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.string)).isRequired,
  formType: PropTypes.string,
  errorMessages: PropTypes.arrayOf(PropTypes.string),
  success: PropTypes.bool,
  messageTitle: PropTypes.string,
  loading: PropTypes.bool,
  links: PropTypes.arrayOf(PropTypes.object),
};
AuthForm.defaultProps = {
  formType: 'Sign Up',
  errorMessages: [],
  success: true,
  messageTitle: '',
  loading: false,
  links: [],
};
export default AuthForm;
