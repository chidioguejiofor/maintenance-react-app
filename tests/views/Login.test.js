import React from 'react';
import { shallow } from 'enzyme';
import chai from 'chai';
import { Login, mapStateToProps,
  mapDispatchToProps } from '../../src/views/auth/Login';
import AuthForm from '../../src/components/Form';

const getMounted = (options = {
  login: {
    errors: { errorMessages: [] },
  },
  history: { push: () => {} },
  sendLoginRequest: () => {}
}) => shallow(<Login {...options} />);
describe('Testing Login Component', () => {
  describe('the components when inputs property are empty rendered in the DOM', () => {
    it('should match snapshot', () => {
      const mountedObj = getMounted();
      expect(mountedObj).toMatchSnapshot();
    });

    it('should render an AuthForm', () => {
      const mountedObj = getMounted();
      chai.expect(mountedObj.find(AuthForm).length)
        .to.equal(1);
      chai.expect(
        mountedObj.find(AuthForm)
          .prop('inputs')
          .length
      )
        .to.equal(2);
    });
  });

  describe('mapStateToProps', () => {
    it('should map the state correctly', () => {
      const mockState = {
        login: 'mockStateLogin',
      };

      const mockedProp = mapStateToProps(mockState);

      chai.expect(mockedProp.login)
        .to.equal(mockState.login);
    });
  });

  describe('mapDispatchToPrps', () => {
    it('should map the state correctly', () => {
      const mockDispatch = jest.fn();

      const mockedProp = mapDispatchToProps(mockDispatch);

      chai.expect(mockedProp.sendLoginRequest)
        .to.be.a('function');
      mockedProp.sendLoginRequest();
      expect(mockDispatch)
        .toHaveBeenCalledTimes(1);
    });
  });
  describe('the methods in the component', () => {
    describe('componentWillReceiveProps', () => {
      it('should  be called when props are updated', () => {
        const spy = jest.spyOn(Login.prototype, 'componentWillReceiveProps');
        const mountedObj = getMounted();
        const newProps = { temp: 'temp Props' };
        mountedObj.setProps(newProps);
        expect(spy)
          .toHaveBeenCalledTimes(1);
      });
      it('should call history.push when success equals true', () => {
        const spy = jest.spyOn(Login.prototype, 'componentWillReceiveProps');
        const historySpy = jest.fn();
        jest.useFakeTimers();
        const newProps = {
          login: {
            errors: { errorMessages: [] },
            success: true
          },
          history: { push: historySpy },
        };
        const options = {
          login: {
            errors: { errorMessages: [] },
          },
          history: { push: historySpy },
        };
        const mountedObj = getMounted(options);

        mountedObj.setProps(newProps);
        expect(spy)
          .toHaveBeenCalledTimes(2);
        expect(setTimeout).toHaveBeenCalledTimes(1);
        jest.runAllTimers();
        expect(historySpy).toHaveBeenNthCalledWith(1, '/dashboard');
      });
    });
    describe('handleChange', () => {
      it('should update state when called', () => {
        const stateSpy = jest.spyOn(Login.prototype, 'setState');
        const spy = jest.spyOn(Login.prototype, 'handleChange');
        const mountedObj = getMounted();

        mountedObj
          .instance()
          .handleChange(null, { name: 'username', value: 'Test Username' });
        expect(spy)
          .toHaveBeenCalledTimes(1);
        expect(stateSpy)
          .toHaveBeenCalledTimes(1);
      });
    });
    describe('handleSubmit', () => {
      it('should call send login request with the an object', () => {
        const sendLoginSpy = jest.fn();
        const spy = jest.spyOn(Login.prototype, 'handleChange');
        const mountedObj = getMounted({
          login: {
            errors: { errorMessages: [] },
          },
          sendLoginRequest: sendLoginSpy,
        });

        mountedObj.instance().handleSubmit();
        expect(sendLoginSpy)
          .toHaveBeenCalledTimes(1);
        expect(spy)
          .toHaveBeenCalledTimes(1);
      });
    });
  });
});
