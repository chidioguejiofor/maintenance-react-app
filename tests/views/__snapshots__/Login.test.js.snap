// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login Component the components when inputs property are empty rendered in the DOM should match snapshot 1`] = `
<div>
  <Header />
  <AuthForm
    errorMessages={Array []}
    formTitle="Login"
    handleChange={[Function]}
    handleSubmit={[Function]}
    iconUrl=""
    inputs={
      Array [
        Object {
          "name": "username",
          "placeholder": "Username",
          "value": "",
        },
        Object {
          "name": "password",
          "placeholder": "Password",
          "type": "password",
          "value": "",
        },
      ]
    }
    links={
      Array [
        Object {
          "caption": "Sign Up",
          "to": "/signup",
        },
      ]
    }
    loading={false}
    messageTitle="Login Successfully"
    submitLoading={false}
    success={true}
    type="auth"
  />
</div>
`;
